File: collectors/weibo.py
"""Weibo hot search collector - web scraping."""
import requests
from bs4 import BeautifulSoup
from typing import List
from .base import BaseCollector, NewsItem


class WeiboCollector(BaseCollector):
    """Fetch Weibo hot search topics."""
    
    HOT_URL = "https://s.weibo.com/top/summary?cate=realtimehot"
    
    def collect(self) -> List[NewsItem]:
        if not self.is_available():
            return []
        
        items = []
        headers = {
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        }
        
        try:
            resp = requests.get(self.HOT_URL, headers=headers, timeout=20)
            resp.encoding = "utf-8"
            soup = BeautifulSoup(resp.text, "html.parser")
            
            # Find the hot search table
            table = soup.find("table")
            if not table:
                return items
            
            rows = table.find_all("tr")[1:]  # Skip header
            for i, row in enumerate(rows[:30]):  # Top 30
                try:
                    td = row.find("td", class_="td-02")
                    if not td:
                        continue
                    
                    a_tag = td.find("a")
                    if not a_tag:
                        continue
                    
                    title = a_tag.get_text(strip=True)
                    href = a_tag.get("href", "")
                    if href.startswith("/"):
                        url = f"https://s.weibo.com{href}"
                    else:
                        url = href
                    
                    # Hot score based on rank
                    hot_score = max(1000 - i * 30, 100)
                    
                    items.append(NewsItem(
                        title=title,
                        url=url,
                        source="weibo",
                        source_name="微博热搜",
                        hot_score=round(hot_score, 1),
                        category="general",
                        language="zh",
                        raw_data={"rank": i + 1}
                    ))
                except Exception:
                    continue
                    
        except Exception as e:
            print(f"[Weibo] Error: {e}")
        
        return items

File: lib/collectors/weibo.ts
import type { Collector } from "./base";
import { createNewsItem } from "./base";

const HOT_URL = "https://s.weibo.com/top/summary?cate=realtimehot";
const MAX_ITEMS = 30;

const TECH_KEYWORDS = [
  "AI", "人工智能", "科技", "手机", "芯片", "互联网",
  "软件", "APP", "数码", "电脑", "特斯拉", "苹果",
  "华为", "微软", "谷歌", "OpenAI", "大模型",
];

export class WeiboCollector implements Collector {
  name = "weibo";
  private config: Record<string, unknown>;

  constructor(config: Record<string, unknown>) {
    this.config = config;
  }

  async collect() {
    const items = [];

    try {
      const headers: Record<string, string> = {
        "User-Agent":
          "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
      };

      const resp = await fetch(HOT_URL, { headers, signal: AbortSignal.timeout(20_000) });
      if (!resp.ok) {
        console.error(`[Weibo] HTTP ${resp.status}`);
        return [];
      }

      const html = await resp.text();

      // Use cheerio for HTML parsing
      let cheerio: any;
      try {
        const mod = await import("cheerio");
        // Cheerio in ESM exports its load function directly
        cheerio = (mod as any).load ? mod : (mod as any).default;
      } catch {
        console.error("[Weibo] cheerio package not available. Install with: npm install cheerio");
        return [];
      }

      const $ = cheerio.load(html);

      // Weibo hot search table rows
      const rows = $("table tr").slice(1); // skip header

      for (let i = 0; i < Math.min(rows.length, MAX_ITEMS); i++) {
        try {
          const row = $(rows[i]);
          const td = row.find("td.td-02");
          if (!td.length) continue;

          const aTag = td.find("a");
          if (!aTag.length) continue;

          const title = aTag.text().trim();
          let href = aTag.attr("href") ?? "";
          if (href.startsWith("/")) {
            href = `https://s.weibo.com${href}`;
          }

          if (!title) continue;

          // Categorize: tech if keyword match
          const isTech = TECH_KEYWORDS.some((kw) => title.includes(kw));
          const hotScore = Math.max(1000 - i * 30, 100);

          items.push(
            createNewsItem(title, href, "weibo", "微博热搜", {
              hot_score: hotScore,
              category: isTech ? "technology" : "general",
              language: "zh",
              raw_data: { rank: i + 1 },
            })
          );
        } catch {
          // per-item error: skip
        }
      }
    } catch (err) {
      console.error(`[Weibo] Error: ${err}`);
    }

    return items;
  }
}
